### ISSUE TITLE
Assertion failed: [10054] in efd_unsignal

### ISSUE THREAD

**Issue by wcyangg** [nanomsg/nanomsg#933]
Environment:
C++ xxx.exe use nanomsg.lib
nanomsg.lib compiled by VS2013 on win10.
xxx.exe run Windows Server 2008 R2

Reproduce:
nn_device module running long time(about 2 hours ?).

Description:
Assert :  [10054] (e:\nanomsg-1.1.2\src\utils\efd_win.inc:265)
look like this problem is also occur in ZeroMQ : 
https://github.com/zeromq/libzmq/issues/1808

**Comment by wcyangg** [comment]
nanomsg version : 1.1.2

**Comment by SG7** [comment]
@wcyangg If you post a test program I can try it on Win10 with VS2015.

**Comment by wcyangg** [comment]
@SG7 
`

char* REPAddress = "tcp://127.0.0.1:45010";
char* REQAddress = "inproc://aTest"

void deviceThread()
{
	int rc;
	int devf;
	int devg;

	/*  Intialise the device sockets. */
	devf = nn_socket(AF_SP_RAW, NN_REP);
	nn_bind(devf, REPAddress);
	devg = nn_socket(AF_SP_RAW, NN_REQ);
	nn_bind(devg, REQAddress);

	/*  Run the device. */
	rc = nn_device(devf, devg);

	/*  Clean up. */
	nn_close(devg);
	nn_close(devf);
}

void workThread()
{
	int sock;
	int endpointId;
	sock = nn_socket(AF_SP, NN_REP);
	endpointId = nn_connect(sock, REQAddress);
	while (server->m_Running) {
		char* buf = (char*)nn_allocmsg(128, 0);
		int bytes = nn_recv(sock, &buf, NN_MSG, 0);
		if (bytes <= 0) {
			break;
		}
		nn_send(sock, buf, strlen(buf), 0);
		nn_freemsg(buf);
	}
	nn_shutdown(sock, endpointId);
	nn_close(sock);
}

void Start(int workCount)
{
	thread dev(deviceThread, this);
	for (int i = 0; i < workCount; i++)
	{
		thread work(workThread, this);
		work.detach();
	}
	dev.join();
}
`

**Comment by gdamore** [comment]
I am thinking that this is suspicious:

```
nn_send(sock, buf, strlen(buf), 0);
nn_freemsg(buf);
```

I'd rather see you pass NN_MSG as a flag to nn_send, and let nn_send release the message.

Having said that, I believe that nn_send should work fine the way you've written it, just with extra data copies.

You *do* have a different problem there though, and that is that you are assuming that the message is ZERO terminated, and yet your reply indicates that you are not sending the terminating NUL byte.  E.g. You need to use strlen(buf) + 1 if you want the zero byte included.

(If the zero byte was not part of the message you received, you might find that strlen does something very unexpected, giving an impossibly large length.

**Comment by gdamore** [comment]
Having said all that, I'm skeptical that this could be the source of any problem.  We're getting an error from Windows recv() function, and the error isn't one we expect.  I'd sure like to know the actual value of that error code.  Can you put a printf() in the efd_win.inc file just above that line, that prints the error code?

**Comment by gdamore** [comment]
BTW, as a workaround, you can *probably* just comment out that assertion check ... it *should* be harmless, but I really want to understand why our assumptions there are wrong.

**Comment by wcyangg** [comment]
@gdamore 
some examples in source code use "test_send" function like this : 
`test_send(endg, "REPLYXYZ");`
`#define test_send(s, d) test_send_impl (__FILE__, __LINE__, (s), (d))
static void NN_UNUSED test_send_impl (char *file, int line,
    int sock, char *data)
{
    size_t data_len;
    int rc;

    **data_len = strlen (data);**

    **rc = nn_send (sock, data, data_len, 0);**
    if (rc < 0) {
        fprintf (stderr, "Failed to send: %s [%d] (%s:%d)\n",
            nn_err_strerror (errno),
            (int) errno, file, line);
        nn_err_abort ();
    }
    if (rc != (int)data_len) {
        fprintf (stderr, "Data to send is truncated: %d != %d (%s:%d)\n",
            rc, (int) data_len,
            file, line);
        nn_err_abort ();
    }
}`

**Comment by skyformat99** [comment]
@wcyangg 
question point   ＞＞   【that prints the error code?】。
so， Please add rc return value。

**Comment by wcyangg** [comment]
@skyformat99 
internal assert fail. no rc value.
windows GetLastError() return 10054

**Comment by gdamore** [comment]
So 10054 is the same as ECONNRESET.  The remote side has reset the connection.

**Comment by gdamore** [comment]
This is the file descriptor I use for signaling.  Probably we can just ignore it.  I'd remove the assert.
